uarray/_uarray_dispatch.cxx

```diff
@@ -0,0 +1,479 @@
+	if (PyTuple_Size(res) != 2)
+		return {};
+
+	auto new_args = py_ref::ref(PyTuple_GET_ITEM(res.get(), 0));
+	auto new_kwargs = py_ref::ref(PyTuple_GET_ITEM(res.get(), 1));
+
+	new_kwargs = canonicalize_kwargs(new_kwargs);
+
+	if (!PyTuple_Check(new_args) || !PyDict_Check(new_kwargs))
+	{
+		PyErr_SetString(PyExc_ValueError, "Invalid return from argument_replacer");
+		return {};
+	}
+
+	return {std::move(new_args), std::move(new_kwargs)};
+}
+
+
+PyObject * uarray_function_call(
+	PyObject * self, PyObject * args, PyObject * kwargs)
+{
+	return reinterpret_cast<Function *>(self)->call(args, kwargs);
+}
+
+
+static py_ref BackendNotImplementedError;
+
+
+PyObject * Function::call(PyObject * args_, PyObject * kwargs_)
+{
+	auto args = canonicalize_args(args_);
+	auto kwargs = canonicalize_kwargs(kwargs_);
+
+	auto backends = py_ref::steal(
+		PyObject_Call(backends_, py_make_tuple(), nullptr));
+	if (!backends)
+		return nullptr;
+
+	auto i_backends = py_ref::steal(PyObject_GetIter(backends));
+	if (!i_backends)
+	{
+		PyErr_SetString(PyExc_TypeError, "backends must return an iterable");
+		return nullptr;
+	}
+
+	while (auto be = py_ref::steal(PyIter_Next(i_backends)))
+	{
+		auto backend = py_ref::steal(PyObject_GetAttrString(be, "backend"));
+		auto coerce = py_ref::steal(PyObject_GetAttrString(be, "coerce"));
+
+		if (!backend || !coerce)
+		{
+			PyErr_SetString(PyExc_TypeError, "Invalid backend spec");
+			return nullptr;
+		}
+
+		if (!PyBool_Check(coerce))
+		{
+			PyErr_SetString(PyExc_TypeError, "coerce must be a Bool");
+			return nullptr;
+		}
+
+		auto new_args = replace_dispatchables(backend, args, kwargs, coerce);
+		if (new_args.args == Py_NotImplemented)
+		{
+			continue;
+		}
+		if (new_args.args == nullptr)
+		{
+			return nullptr;
+		}
+
+		auto ua_function =
+			py_ref::steal(PyObject_GetAttrString(backend, "__ua_function__"));
+		if (!ua_function)
+		{
+			PyErr_SetString(PyExc_TypeError, "Backend must have __ua_function__");
+			return nullptr;
+		}
+
+		auto ua_func_args = py_make_tuple(
+			reinterpret_cast<PyObject *>(this), new_args.args, new_args.kwargs);
+		auto result = py_ref::steal(
+			PyObject_Call(ua_function, ua_func_args, nullptr));
+		if (result != Py_NotImplemented)
+			return result.release();
+	}
+
+	if (def_impl_ == Py_None)
+	{
+		PyErr_SetString(
+			BackendNotImplementedError,
+			"No selected backends had an implementation for this function.");
+		return nullptr;
+	}
+
+	return PyObject_Call(def_impl_, args, kwargs);
+}
+
+
+PyObject * Function_repr(Function * self)
+{
+	if (self->dict_)
+		if (auto name = PyDict_GetItemString(self->dict_, "__name__"))
+			return PyUnicode_FromFormat("<uarray multimethod '%S'>", name);
+
+	return PyUnicode_FromString("<uarray multimethod>");
+}
+
+
+/** Implements the descriptor protocol to allow binding to class instances */
+PyObject * Function_descr_get(PyObject * self, PyObject * obj, PyObject * type)
+{
+	Py_INCREF(self);
+	if (!obj)
```

Comment:
You might want to compare to `Py_None` here.

Reply:
If I call `func.__get__(None)` this receives `nullptr`.

Reply:
Probably need a `Py_INCREF` for `obj` here.